components/plurals/src/provider.rs

```diff
@@ -368,115 +377,705 @@ pub enum PluralElementsKeysV1 {
+        // Safety: by invariant, `self.0` contains valid values for PluralCategoryAndMetadata
+        unsafe { PluralCategoryAndMetadata::try_from_unpacked(unpacked).unwrap_unchecked() }
+    }
 }
 
-impl<'a> PluralElementsV1<'a> {
-    /// Returns the string for the given [`PluralOperands`] and [`PluralRules`].
-    pub fn get(&'a self, op: PluralOperands, rules: &PluralRules) -> &'a str {
+impl PluralCategoryAndMetadata {
+    fn try_from_unpacked(unpacked: PluralCategoryAndMetadataUnpacked) -> Option<Self> {
+        let plural_category = PluralElementsKeysV1::new_from_u8(unpacked.plural_category_byte)?;
+        let metadata = FourBitMetadata::try_from_byte(unpacked.metadata_byte)?;
+        Some(Self {
+            plural_category,
+            metadata,
+        })
+    }
+}
+
+impl AsULE for PluralCategoryAndMetadata {
+    type ULE = PluralCategoryAndMetadataPackedULE;
+    #[inline]
+    fn to_unaligned(self) -> Self::ULE {
+        PluralCategoryAndMetadataPackedULE::from(self)
+    }
+    #[inline]
+    fn from_unaligned(unaligned: Self::ULE) -> Self {
+        unaligned.get()
+    }
+}
+
+/// A bitpacked DST for [`PluralElements`].
+///
+/// Can be put in a [`Cow`] or a [`VarZeroVec`].
+#[derive(Debug, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct PluralElementsPackedULE<V: VarULE + ?Sized> {
+    _v: PhantomData<V>,
+    /// Invariant Representation:
+    ///
+    /// First byte: `d...mmmm`
+    /// - `d` = 0 if singleton, 1 if a map
+    /// - `...` = padding, should be 0
+    /// - `mmmm` = [`FourBitMetadata`] for the default value
+    ///
+    /// If d is 0:
+    /// - Remainder: the default (plural "other") value `V`
+    ///
+    /// If d is 1:
+    /// - Second byte: L = the length of `V`
+    /// - Bytes 2..(2+L): the default (plural "other") value `V`
+    /// - Remainder: [`PluralElementsTupleSliceVarULE`]
+    bytes: [u8],
+}
+
+/// The type of the special patterns list.
+type PluralElementsTupleSliceVarULE<V> = VarZeroSlice<VarTupleULE<PluralCategoryAndMetadata, V>>;
+
+/// The type of the default value.
+type PluralElementWithMetadata<'a, T> = (FourBitMetadata, &'a T);
+
+/// Internal intermediate type that can be converted into a [`PluralElementsPackedULE`].
+struct PluralElementsPackedBuilder<'a, T> {
+    pub default: PluralElementWithMetadata<'a, T>,
+    pub specials: Option<Vec<VarTuple<PluralCategoryAndMetadata, &'a T>>>,
+}
+
+/// Internal unpacked and deserialized values from a [`PluralElementsPackedULE`].
+struct PluralElementsUnpacked<'a, V: VarULE + ?Sized> {
+    pub default: PluralElementWithMetadata<'a, V>,
+    pub specials: Option<&'a PluralElementsTupleSliceVarULE<V>>,
+}
+
+/// Internal unpacked bytes from a [`PluralElementsPackedULE`].
+struct PluralElementsUnpackedBytes<'a> {
+    pub lead_byte: u8,
+    pub v_bytes: &'a [u8],
+    pub specials_bytes: Option<&'a [u8]>,
+}
+
+/// Helper function to access a value from [`PluralElementsTupleSliceVarULE`]
+fn get_special<V: VarULE + ?Sized>(
```

Comment:
nit: this can be in `get`

Reply:
I see this as an associated function on `type PluralElementsTupleSliceVarULE<V>`, but I can't add associated functions to typedefs. I also stylistically prefer to keep functions at the module level because there is less nesting / indentation.

But, I have the sense that you feel more strongly about this style (fully encapsulating private functions and structs when possible) than I do, so I'll move it for you.

Reply:
Actually, I ended up calling this function in one more place (the new serde impl) so I'm leaving it where it is.